tree/ntuple/v7/doc/architecture.md

```diff
@@ -3,7 +3,439 @@ RNTuple Code Architecture
 
 > This document is meant for ROOT developers. It provides background information on the RNTuple code design and behavior.
 
-> This document is currently a stub.
+> The RNTuple code uses the nomenclature from the RNTuple format specification (e.g. "field", "column", "anchor", etc.).
+
+General Principles
+------------------
+
+The RNTuple classes provide the functionality to read, write, and describe RNTuple datasets.
+The core classes, such as `RNTupleReader` and `RNTupleWriter`, are part of the RNTuple library.
+Additional tools, such as the `RNTupleImporter` and the `RNTupleInspector`, are part of the RNTupleUtils library,
+that depends on the RNTuple library.
+
+The RNTuple classes are organized in layers:
+the storage layer, the primitives layer, the logical layer and the event iteration.
+Most classes in the storage layer and the primitives layer are in the `ROOT::Internal` namespace,
+with the notable exception of the descriptor classes (`RNTupleDescriptor`, `RFieldDescriptor`, etc.).
+Most classes in the upper layers provide public interfaces.
+
+| Layer      | Description                                                         | Example of classes                                   |
+|------------|---------------------------------------------------------------------|------------------------------------------------------|
+| Storage    | Read and write pages (physical: file, DAOS; virtual: e.g. buffered) | RPage{Source,Sink}, RNTupleDescriptor, RClusterPool  |
+| Primitives | Storage-backed columns of simple types                              | RColumn, RColumnElement, RPage                       |
+| Logical    | Mapping of C++ types onto columns                                   | RField, RNTupleModel, REntry                         |
+| Iteration  | Reading and writing events / properties                             | RNTuple{Reader,Writer}, RNTupleView, RNTupleDS (RDF) |
+| Tooling    | Higher-level, RNTuple related utility classes                       | RNTupleMerger, RNTupleImporter, RNTupleInspector     |
+
+The RNTuple classes are, unless explicitly stated otherwise, conditionally thread safe.
+
+The read and write APIs provide templated, compile-time type-safe APIs,
+APIs where the type at hand is passed as string and which are runtime type-safe,
+and type-unsafe APIs using void pointers.
+
+On I/O errors and invalid input, RNTuple classes throw an `RException`.
```

Comment:
This is more for future reference/note to self, but we should check if this is the case everywhere (IIRC there still are a few places, accessible through public APIs, with `asserts`).